godal.go

```diff
@@ -3923,6 +3923,133 @@ func (ds *Dataset) NearblackInto(sourceDs *Dataset, switches []string, opts ...N
+	defer cIds.free()
+	cInfos := sliceToCStringArray(infos)
+	defer cInfos.free()
+
+	return C.godalGCPPropertiesToGDALGCP(C.int(len(GCPList)), cIds.cPointer(), cInfos.cPointer(), cDoubleArray(gcpPixels), cDoubleArray(gcpLines), cDoubleArray(gcpXs), cDoubleArray(gcpYs), cDoubleArray(gcpZs))
+}
+
+// gdalGCPToGoGCPArray is a utility function for conversion from `*C.GDAL_GCP` (GDAL) to `[]GCP` (Go)
+func gdalGCPToGoGCPArray(gcp *C.GDAL_GCP, numGCPs int) []GCP {
+	var ret = make([]GCP, numGCPs)
+
+	//https://github.com/golang/go/wiki/cgo#turning-c-arrays-into-go-slices
+	gcps := (*[1 << 30]C.GDAL_GCP)(unsafe.Pointer(gcp))
+	for i := 0; i < numGCPs; i++ {
+		ret[i] = GCP{
+			pszId:      C.GoString(gcps[i].pszId),
+			pszInfo:    C.GoString(gcps[i].pszInfo),
+			dfGCPPixel: float64(gcps[i].dfGCPPixel),
+			dfGCPLine:  float64(gcps[i].dfGCPLine),
+			dfGCPX:     float64(gcps[i].dfGCPX),
+			dfGCPY:     float64(gcps[i].dfGCPY),
+			dfGCPZ:     float64(gcps[i].dfGCPZ),
+		}
+	}
+	return ret
+}
+
+// GetGCPSpatialRef runs the GDALGetGCPSpatialRef function
+func (ds *Dataset) GetGCPSpatialRef() *SpatialRef {
+	return &SpatialRef{handle: C.godalGetGCPSpatialRef(ds.handle()), isOwned: false}
+}
+
+// GetGCPCount runs the GDALGetGCPCount function
+func (ds *Dataset) GetGCPCount() int {
+	return int(C.godalGetGCPCount(ds.handle()))
+}
+
+// GetGCPs runs the GDALGetGCPs function
+// TODO: This makes 2 cgo calls, we could reduce it to 1 by combining `godalGetGCPs` and `GetGCPCount`
+func (ds *Dataset) GetGCPs() []GCP {
+	hndl := C.godalGetGCPs(ds.handle())
+	numGCPs := ds.GetGCPCount()
+	if hndl != nil {
+		return gdalGCPToGoGCPArray(hndl, numGCPs)
+	}
+	return []GCP{}
+}
+
+// GetGCPProjection runs the GDALGetGCPProjection function
+func (ds *Dataset) GetGCPProjection() string {
+	return C.GoString(C.godalGetGCPProjection(ds.handle()))
+}
+
+// SetGCPs runs the GDALSetGCPs function
+func (ds *Dataset) SetGCPs(GCPList []GCP, pszGCPProjection string, opts ...SetGCPsOption) error {
+	setGCPsOpts := setGCPsOpts{}
+	for _, opt := range opts {
+		opt.setSetGCPsOpt(&setGCPsOpts)
+	}
+	cgc := createCGOContext(nil, setGCPsOpts.errorHandler)
+
+	GCPProj := C.CString(pszGCPProjection)
+	defer C.free(unsafe.Pointer(GCPProj))
+
+	C.godalSetGCPs(cgc.cPointer(), ds.handle(), C.int(len(GCPList)), goGCPArrayToGDALGCP(GCPList), GCPProj)
+	if err := cgc.close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// SetGCPs2 runs the GDALSetGCPs2 function
+func (ds *Dataset) SetGCPs2(GCPList []GCP, sr *SpatialRef, opts ...SetGCPsOption) error {
+	setGCPsOpts := setGCPsOpts{}
+	for _, opt := range opts {
+		opt.setSetGCPsOpt(&setGCPsOpts)
+	}
+	cgc := createCGOContext(nil, setGCPsOpts.errorHandler)
+
+	srHandle := C.OGRSpatialReferenceH(nil)
+	if sr != nil {
+		srHandle = sr.handle
+	}
+
+	C.godalSetGCPs2(cgc.cPointer(), ds.handle(), C.int(len(GCPList)), goGCPArrayToGDALGCP(GCPList), srHandle)
+	if err := cgc.close(); err != nil {
+		return err
+	}
+	return nil
+}
```

Comment:
Given that gdal's setgcp call accepts an empty projectionstring or projectionref, I would suggest that you use only a single golang SetGCPs function
```go func (ds *Dataset) SetGCPs(GCPList []GCP,  opts ...SetGCPsOption) error```
and allow passing the projection string or ref through SetGCPOptions.
i.e. as a user you can call either:

- `ds.SetGCPs(mygcplist)`
- `ds.SetGCPs(mygcplist, godal.GCPProjection("+proj=lonlat +foo=bar"))`
- `ds.SetGCPs(mygcplist, godal.GCPSpatialRef(sr) //sr is a *godal.SpatialRef`